File: src/rasterize/fastRasterizer.py
import numpy as np
from ...lib.rgb import hexToRGB

def blendOver(canvas: np.ndarray, mask: np.ndarray, colorRGB, opacity: float) -> None: # Porter-Duff
    if opacity <= 0:
        return

    idx = mask.astype(bool)
    if not idx.any():
        return

    srcA = float(opacity)
    dstA = canvas[idx, 3].astype(np.float32) / 255.0
    outA = srcA + dstA * (1.0 - srcA)

    srcRGB = np.asarray(colorRGB, dtype=np.float32)
    dstRGB = canvas[idx, :3].astype(np.float32)

    safe = outA > 1e-6
    blended = srcRGB * srcA + dstRGB * dstA[:, None] * (1.0 - srcA)
    outRGB = np.zeros_like(blended)
    outRGB[safe] = blended[safe] / outA[safe, None]

    canvas[idx, :3] = np.clip(outRGB, 0, 255).astype(np.uint8)
    canvas[idx, 3] = np.clip(outA * 255.0, 0, 255).astype(np.uint8)


def clipBB(bbox, height, width):
    x0, y0, x1, y1 = bbox
    x0, y0 = max(0, int(x0)), max(0, int(y0))
    x1, y1 = min(width, int(x1)), min(height, int(y1))
    if x1 <= x0 or y1 <= y0:
        return None
    return x0, y0, x1, y1


def rasterizeShape(shape, canvas: np.ndarray, height: int, width: int) -> None:
    clipped = clipBB(shape.boundingBox(), height, width)
    if clipped is None:
        return

    x0, y0, x1, y1 = clipped
    region = canvas[y0:y1, x0:x1]

    if getattr(shape, "inner", None) is not None:
        mask = np.zeros((y1 - y0, x1 - x0), dtype=np.uint8)
        shape.paintOnMask(mask, filled=True, origin=(x0, y0))
        blendOver(region, mask, hexToRGB(shape.inner.color), shape.inner.opacity)

    if getattr(shape, "outer", None) is not None:
        mask = np.zeros((y1 - y0, x1 - x0), dtype=np.uint8)
        shape.paintOnMask(mask, filled=False, origin=(x0, y0))
        blendOver(region, mask, hexToRGB(shape.outer.color), shape.outer.opacity)


def rasterize(svg, height: int, width: int) -> np.ndarray:
    canvas = np.zeros((height, width, 4), dtype=np.uint8)
    for elm in svg.data:
        rasterizeShape(elm, canvas, height, width)
    return canvas

File: src/rasterize/fastRasterizer.cc
#include "fastRasterizer.hh"

#include <algorithm>
#include <cmath>

#include "../../lib/rgb/rgb.hh"

void blendOver(cv::Mat& canvas, const cv::Mat& mask, const cv::Vec3f& colorRGB, double opacity) {
    if (opacity <= 0.0) return;
    if (cv::countNonZero(mask) == 0) return;

    CV_Assert(canvas.type() == CV_8UC4);
    CV_Assert(mask.type() == CV_8UC1);
    CV_Assert(canvas.size() == mask.size());

    const double srcA = opacity;

    for (int y = 0; y < canvas.rows; ++y) {
        for (int x = 0; x < canvas.cols; ++x) {
            if (!mask.at<uint8_t>(y, x)) continue;

            cv::Vec4b& px = canvas.at<cv::Vec4b>(y, x);

            const double dstA = px[3] / 255.0;
            const double outA = srcA + dstA * (1.0 - srcA);

            const cv::Vec3d dstRGB(px[0], px[1], px[2]);
            const cv::Vec3d srcRGB(colorRGB[0], colorRGB[1], colorRGB[2]);
            const cv::Vec3d blended = srcRGB * srcA + dstRGB * dstA * (1.0 - srcA);

            cv::Vec3d outRGB(0.0, 0.0, 0.0);
            if (outA > 1e-6) {
                outRGB = blended / outA;
            }

            px[0] = static_cast<uint8_t>(std::clamp(outRGB[0], 0.0, 255.0));
            px[1] = static_cast<uint8_t>(std::clamp(outRGB[1], 0.0, 255.0));
            px[2] = static_cast<uint8_t>(std::clamp(outRGB[2], 0.0, 255.0));
            px[3] = static_cast<uint8_t>(std::clamp(outA * 255.0, 0.0, 255.0));
        }
    }
}

std::optional<cv::Rect> clipBB(const lib::BBox& bbox, int height, int width) {
    int x0 = std::max(0, static_cast<int>(bbox.x0));
    int y0 = std::max(0, static_cast<int>(bbox.y0));
    int x1 = std::min(width, static_cast<int>(bbox.x1));
    int y1 = std::min(height, static_cast<int>(bbox.y1));

    if (x1 <= x0 || y1 <= y0) return std::nullopt;
    return cv::Rect(x0, y0, x1 - x0, y1 - y0);
}

void rasterizeShape(const Shape& shape, cv::Mat& canvas, int height, int width) {
    auto clipped = clipBB(shape.boundingBox(), height, width);
    if (!clipped) return;

    cv::Rect region = *clipped;
    cv::Mat view = canvas(region);

    if (shape.inner().has_value()) {
        cv::Mat mask = cv::Mat::zeros(region.height, region.width, CV_8UC1);
        shape.paintOnMask(mask, /*filled=*/true, region.tl());
        blendOver(view, mask, lib::hexToRGB(shape.inner()->color), shape.inner()->opacity);
    }

    if (shape.outer().has_value()) {
        cv::Mat mask = cv::Mat::zeros(region.height, region.width, CV_8UC1);
        shape.paintOnMask(mask, false, region.tl());
        blendOver(view, mask, lib::hexToRGB(shape.outer()->color), shape.outer()->opacity);
    }
}

cv::Mat rasterize(const SVG& svg, size_t height, size_t width) {
    cv::Mat canvas = cv::Mat::zeros(height, width, CV_8UC4);
    for (const auto& elm : svg.data()) {
        rasterizeShape(*elm, canvas, height, width);
    }
    return canvas;
}